python/shotgun_desktop/startup.py

```diff
@@ -698,6 +699,45 @@ def get_bundle_cache_location(self):
             return os.environ.get("SHOTGUN_DESKTOP_BUNDLE_CACHE_LOCATION") or None
 
 
+def automatic_claims_renewal(user, preemtive_renewal_threshold=0.9):
+    """
+    Handles automatic renewal of the SAML2 claims for the user.
+
+    :params user: an already logged in user.
+    :params preemtive_renewal_threshold: How far into the claims duration we will attempt renewal.
+                                         Defaults to 90%, usually 3 minutes 45 seconds.
+
+    This will only be used in a context where SSO is enabled on the
+    Shotgun site.
+    """
+    from tank.authentication import interactive_authentication
+    from sgtk.authentication import AuthenticationCancelled
+
+    logger.debug("Automatic claims renewal")
+    try:
+        previous_expiration = user._impl.get_saml_expiration()
```

Comment:
That right there shows that this functionality should probably be in core or some refactoring of how this can be exposed. You're accessing the internal API of the user class.